### PR TITLE
analyze: fix timestamp

### LINKED ISSUE #34670: Time zone rejected in timestamps
### systemd version the issue has been seen with

256.6-1

### Used distribution

ArchLinux

### Linux kernel version used

6.11.2-arch1-1

### CPU architectures issue was seen on

x86_64

### Component

systemctl, systemd-analyze

### Expected behaviour you didn't see

The manual page for timestamps (section *parsing timestamps*) cites `Fri 2012-11-23 23:02:15 CET` as an example of valid timestamp in the first paragraph, and tells in the second one that time zones "*may be given after a space [...] it can be: "UTC", an entry in the installed IANA timezone database ("CET", [...])*".

The end of the section tells timestamps can be validated with `systemd-analyze timestamp`. Hence, the reader would expect that the following command validates the timestamp string.
```
$ systemd-analyze timestamp 'Fri 2012-11-23 23:02:15 CET'
  Original form: Fri 2012-11-23 23:02:15 CET
Normalized form: Fri 2012-11-23 23:02:15 CET
       (in UTC): Fri 2012-11-23 22:02:15 UTC
   UNIX seconds: @1353708135
       From now: 11 years 10 months ago
```

### Unexpected behaviour you saw

However, the command produces this output:
```
$ systemd-analyze timestamp 'Fri 2012-11-23 23:02:15 CET'
Failed to parse "Fri 2012-11-23 23:02:15 CET": Invalid argument
Hint: this expression is a valid calendar specification. Use 'systemd-analyze calendar "Fri 2012-11-23 23:02:15 CET"' instead
```

This timestamp is also rejected in other `systemd` commands. For example,
```
$ systemctl poweroff --when='Fri 2012-11-23 23:02:15 CET'
Failed to parse --when= argument 'Fri 2012-11-23 23:02:15 CET': Invalid argument
```

### Steps to reproduce the problem

Run the commands above.

### Additional program output to the terminal or log subsystem illustrating the issue

_No response_

### PR DESCRIPTION
**Pull request by yuwata**
Fixes #34670.

### CODE REVIEW

**Comment by poettering** [comment]
CI is not happy with this, the tests this change touches fail now!

**Comment by poettering** [comment]
my guess is that Asia/Tokyo might not be installed on the CI?

**Comment by poettering** [comment]
Hmm, I take it time diff of Tokyo to GMT is 9h, right? Hence 'Thu 1970-01-01 09:01 Asia/Tokyo' is 1s after the UNIX epoch?

**Comment by yuwata** [comment]
> my guess is that Asia/Tokyo might not be installed on the CI?

Right. Should be fixed now.

> Hmm, I take it time diff of Tokyo to GMT is 9h, right? Hence 'Thu 1970-01-01 09:01 Asia/Tokyo' is 1s after the UNIX epoch?

Yes.

File: test/units/TEST-65-ANALYZE.sh [hunk 1]
```diff
@@ -172,6 +172,15 @@ systemd-analyze calendar --base-time=yesterday --iterations=5 '*-* *:*:*'
 systemd-analyze timestamp now
 systemd-analyze timestamp -- -1
 systemd-analyze timestamp yesterday now tomorrow
+systemd-analyze timestamp 'Fri 2012-11-23 23:02:15'
+systemd-analyze timestamp 'Fri 2012-11-23 23:02:15 UTC'
+systemd-analyze timestamp 'Fri 2012-11-23 23:02:15 CET'
+for i in $(timedatectl list-timezones); do
+    # FIXME(?): NZ is broken???
+    [[ "$i" == NZ ]] && continue
```
Comment (poettering) [inline test/units/TEST-65-ANALYZE.sh]: this is actually a bug in our parser.

it's the "welded RFC3339" style, that uses "Z" as suffix. i.e. the first if check in parse_timestamp() misinterprets "NZ" suffix as such an ISO timestamp.

I guess the line needs to be changed to:

```c
if (t_len > 2 && t[t_len - 1] == 'Z' && undecchar(t[t_len - 2]) >= 0)  /* RFC3339-style welded UTC: "1985-04-12T23:20:50.52Z" */
```

or so

Comment (poettering) [reply test/units/TEST-65-ANALYZE.sh]: i.e. it's caused by ef658a63f81. Any chance you can add a commit that tweaks this, and mentions it is a follow-up for ef658a63f81?

**Comment by poettering** [comment]
lgtm, but i hope i can convince you to fix the NZ misparse at the same time!

**Comment by yuwata** [comment]
> lgtm, but i hope i can convince you to fix the NZ misparse at the same time!

Done in the first commit. PTAL.

**Comment by poettering** [comment]
excellent! thanks!

**Review by bluca — APPROVED** [review]

**Pull request merged**